sys/include/senml.h

```diff
@@ -0,0 +1,158 @@
+    char               *base_name;       /**< Concatenated with name attribute gives sensor's UID   */
+    double              base_time;       /**< Update time values are relative to this time          */
+    char               *base_unit;       /**< When all values have the same unit, set this field    */
+    double              base_value;      /**< If not zero, add this to (float) value in each record */
+    double              base_sum;        /**< If not zero, add this to sum in each record           */
+} senml_base_info_t;
+
+/**
+ * @brief struct that contains the values of a SenML record
+ */
+typedef struct {
+    char              *name;        /**< Sensor's name, will be appended to base name if provided */
+    char              *unit;        /**< Unit of the measurement                                  */
+    char              *link;        /**< Link to additional information (must be escaped!)        */
+    double             time;        /**< Time (seconds since unix epoch) of the measurement       */
+    double             update_time; /**< Time before sensor provides an updated measurement       */
+    double             value_sum;   /**< Integrated sum of the values over time                   */
+    senml_value_type_t value_type;  /**< Indicates which type the value is of                     */
+    union {
+        double            f;  /**< A float value                              */
+        char             *s;  /**< A string value                             */
+        bool              b;  /**< A boolean value                            */
+        char             *d;  /**< A data value (binary data, base64 encoded) */
+    } value;
+} senml_record_t;
+
+/**
+ * @brief struct that holds a SenML pack (optional base info and 1..n records)
+ */
+typedef struct {
+    senml_base_info_t  *base_info;  /**< Pointer to the base info, may be NULL           */
+    senml_record_t     *records;    /**< Pointer to the beginning of an array of records */
+    size_t              num;        /**< Number of records in the array                  */
+} senml_pack_t;
+
+/**
+ * @brief Decodes a SenML document in JSON format
+ *
+ * @param[in]  input The JSON document containing the SenML pack
+ * @param[out] pack  The @ref senml_pack_t structure in which the
+ *                   records and base info will be stored.
+ *
+ * @return 0 on success, -1 if @p input is not a valid SenML document, or a
+ *         positive number indicating by how much the number of records in
+ *         @p input exceeds `pack->num`
+ */
+int senml_decode_json_s(char *input, senml_pack_t *pack);
+
+/**
+ * @brief Creates a SenML document in JSON format from a SenML pack
+ *
+ * @param[in]  pack   The pack that contains the optional base info and the records
+ * @param[out] output The resulting JSON document
+ * @param[in]  len    The length of @p output in bytes
+ *
+ * @return 0 on success, or -1 when @p output is not big enough
```

Comment:
I would prefer to add named return values, something like
```c
// somewhere in the top of the file
enum {
    SENML_OK     = 0,
    SENML_OVRFLW = -1,
    ...
};

// here
/**
 * @return  SENML_OK on success
 * @return  SENML_OVRFLW if @p output is not large enough
 */
```
Same goes of course for `senml_decode_json_s()`